More flexible authentication header parsing in DispatchingAuthenticator

drlue opened the pull request:
Removed Response.challenges() and replace by more flexible authentication header parsing.
Authentication headers with wrongly ordered parameters not following RFC 2617 were not recognized.
See: https://github.com/square/okhttp/issues/2780

Thanks for your great library!
Hopfefully you find this usefull, or at least it can give you some input for improvement.

regards
Lukas

rburgst commented:
Hi

can you please add a unit test to illustrate where the built-in parsing does not suffice?

thanks
- Rainer

drlue commented:
Hi Rainer,
yes for sure, unfortunately I can't add a test for real life situation, as it is the server of one of my app users.
regards
Lukas

rburgst commented:
well you can simply maintain the structure of the request but mask any personally identifying data.

alexeyvasilyev commented:
Looks good for me.
I'm also using a workaround for OkHttp bug https://github.com/square/okhttp/issues/2780. But it is not so elegant as this code.
However I'm thinking than you need to submit bug fix in OkHttp library as well.

drlue commented:
Hi Rainer, I added the test.
Yes you are right, it would be good to add the bug fix also there. Basically its not bug at all, some servers just don't use the correct ordering…
When I have time I also will propose this fix in okhttp.

regards and have a good week
Lukas

drlue commented:
Hi,
I also added a pull request to okhttp for the bugfix.
I also fixed a bug in the new implementation where wrongly cased realm -> ReAlM was not detected.
regards Lukas

alexeyvasilyev commented:
I guess this commit should be reverted since it is fixed in okhttp 3.5.0.

rburgst commented:
can you check if it works without?

rburgst commented:
ok. Reverted the change and now upgraded to okhttp3.5

drlue commented:
Yes it was fixed, sorry I forgot to link to the okhttp pullrequest https://github.com/square/okhttp/pull/2921.
